pkg/oauth/apiserver/registry/personaltokens/delegate/delegate.go

```diff
@@ -0,0 +1,186 @@
+	if token == nil {
+		return nil, errors.NewInternalError(fmt.Errorf("failed to convert generic accesstoken GET result to its typed version"))
+	}
+
+	if token.UserName != ctxUserName { // don't reveal other people tokens' hashes
+		return nil, errors.NewNotFound(oauth.Resource("personaltoken"), name)
+	}
+
+	return &oauthapi.PersonalAccessToken{OAuthAccessToken: *token}, nil
+}
+
+func (r *REST) Watch(ctx context.Context, options *metainternal.ListOptions) (watch.Interface, error) {
+	if ctx == nil {
+		return nil, fmt.Errorf("context is nil")
+	}
+
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sanitizedListOpts, err := userFieldSelectorToListOptions(options, ctxUserName)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenListWatcher, err := r.accessTokenStorage.Watch(ctx, sanitizedListOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	watcher := newOAuthAccessTokenWatcher(tokenListWatcher.ResultChan())
+	go watcher.Watch()
+	return watcher, nil
+}
+
+func (r *REST) ConvertToTable(ctx context.Context, object runtime.Object, tableOptions runtime.Object) (*metav1.Table, error) {
+	return r.tableConvertor.ConvertToTable(ctx, object, tableOptions)
+}
+
+func (r *REST) NamespaceScoped() bool {
+	return false
+}
+
+func (r *REST) Delete(ctx context.Context, name string, validateFunc rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
+	if _, err := r.Get(ctx, name, &metav1.GetOptions{}); err != nil {
+		return &metav1.Status{Status: metav1.StatusFailure}, false, err
+	}
+
+	return r.accessTokenStorage.Delete(ctx, name, validateFunc, options)
+}
+
+func getUserFromContext(ctx context.Context) (string, error) {
+	userInfo, exists := apirequest.UserFrom(ctx)
+	if !exists {
+		return "", kerrors.NewForbidden(oauth.Resource("personaltokens"), "", fmt.Errorf("unable to get personal tokens without a user in context"))
+	}
+
+	userName := userInfo.GetName()
+	if len(userName) == 0 {
+		return "", kerrors.NewForbidden(oauth.Resource("personaltokens"), "", fmt.Errorf("user has a zero-length username"))
+	}
+
+	return userName, nil
+}
+
+func oauthAccessTokenListToUserOAuthAccessTokenList(l *oauthapi.OAuthAccessTokenList) *oauthapi.PersonalAccessTokenList {
+	ret := oauthapi.PersonalAccessTokenList{}
+	for _, t := range l.Items {
+		ret.Items = append(ret.Items, oauthapi.PersonalAccessToken{OAuthAccessToken: t})
+	}
+	return &ret
+}
+
+func userFieldSelectorToListOptions(opts *metainternal.ListOptions, userName string) (*metainternal.ListOptions, error) {
+	var newOpts *metainternal.ListOptions
+	if opts == nil {
+		newOpts = &metainternal.ListOptions{}
+	} else {
+		newOpts = opts.DeepCopy()
+	}
+
+	// the current field selector cannot make any assumptions over the "userName" field
+	for _, req := range newOpts.FieldSelector.Requirements() {
+		if req.Field == "userName" {
+			return nil, errors.NewUnauthorized("the userName field selector is not allowed")
```

Comment:
can we alternatively just add our userName constraint? If the user adds another one for a different user, the set will just be empty. That's a cleaner and maybe safer solution.

Reply:
You actually do that below. I.e. we could just drop this whole check?